signer/src/bitcoin/validation.rs

```diff
@@ -28,129 +37,396 @@ pub struct BitcoinTxContext {
     /// bitcoin blockchain with the greatest height. On ties, we sort by
     /// the block hash descending and take the first one.
     pub chain_tip: BitcoinBlockHash,
-    /// How many bitcoin blocks back from the chain tip the signer will
-    /// look for requests.
-    pub context_window: u16,
     /// The block height of the bitcoin chain tip identified by the
     /// `chain_tip` field.
     pub chain_tip_height: u64,
-    /// The transaction that is being validated.
-    pub tx: BitcoinTx,
+    /// This contains each of the requests for the entire transaction
+    /// package. Each element in the vector corresponds to the requests
+    /// that will be included in a single bitcoin transaction.
+    pub request_packages: Vec<TxRequestIds>,
+    /// This signer's public key.
+    pub signer_public_key: PublicKey,
+    /// The current aggregate key that was the output of DKG.
+    pub aggregate_key: PublicKey,
+    /// The state of the signers.
+    pub signer_state: SignerBtcState,
+}
+
+/// This type is a container for all deposits and withdrawals that are part
+/// of a transaction package.
+#[derive(Debug, Clone)]
+pub struct TxRequestIds {
     /// The deposit requests associated with the inputs in the transaction.
-    pub deposit_requests: Vec<OutPoint>,
-    /// The total amount of the transaction fee in sats.
-    pub tx_fee: u64,
-    /// The current market fee rate in sat/vByte.
-    pub fee_rate: f64,
-    /// The total fee amount and the fee rate for the last transaction that
-    /// used this UTXO as an input.
-    pub last_fee: Option<Fees>,
+    pub deposits: Vec<OutPoint>,
     /// The withdrawal requests associated with the outputs in the current
     /// transaction.
-    pub request_ids: Vec<QualifiedRequestId>,
-    /// The public key of the signer that created the bitcoin transaction.
-    /// This is very unlikely to ever be used in the
-    /// [`BitcoinTx::validate`] function, but is here for logging and
-    /// tracking purposes.
-    pub origin: PublicKey,
-    /// Two byte prefix for BTC transactions that are related to the Stacks
-    /// blockchain.
-    pub magic_bytes: [u8; 2],
+    pub withdrawals: Vec<QualifiedRequestId>,
+}
+
+/// Check that this does not contain duplicate deposits or withdrawals.
+///
+/// TODO: Implement.
+pub fn is_unique(_packages: &[TxRequestIds]) -> bool {
+    false
 }
 
 impl BitcoinTxContext {
     /// Validate the current bitcoin transaction.
-    pub async fn validate<C>(&self, _ctx: &C) -> Result<(), Error>
+    pub async fn pre_validation<C>(&self, _ctx: &C) -> Result<(), Error>
     where
         C: Context + Send + Sync,
     {
-        unimplemented!()
-    }
+        let no_withdrawals = self
+            .request_packages
+            .iter()
+            .all(|reqs| reqs.withdrawals.is_empty());
```

Comment:
Just wondering why should we assure that there are no withdrawal requests?

Reply:
I thought that it would be easier to understand that withdrawals fail validation with a simple check here. I'm okay with removing it since withdrawals will still fail validation because of the `is_valid_tx` check.

Reply:
I am ok both ways, but I slightly prefer removing it from here if we already have a fail-safe. Just to ensure we don't forget to remove the check from multiple places when we implement the withdrawals.